### PR TITLE
Common terms search

### PR DESCRIPTION
**Pull request by neilbmclaughlin**
I've moved some of the tests around to allow `scripts/test` to be run in the container both locally and in CI.

Due to the reshaping of the `doctors` property this is a breaking change for `gp-finder` which will be resolved by https://github.com/nhsuk/gp-finder/pull/128.

### CODE REVIEW

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-22.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-22.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-22.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-22.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-22.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-22.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-22.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-22.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-22.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-22.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-22.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-22.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-22.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-22.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-22.dev.beta.nhschoices.net)

File: scripts/container-tests [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  else
+    exit_status=1
+    echo "Max file descriptors test failed. Expected $expected_max, got $max_file_descriptors"
+  fi
+}
+
+test_number_of_docs() {
+  echo "Test number of docs in index"
+  expected_docs=9537
+  number_of_docs=$(curl -s "http://es:9200/profiles/_count?q=*" | jq '.count')
```
Comment (neilbmclaughlin) [inline scripts/container-tests:43]: Changed this from the inspection of the es-stats.json file as the introduction of nested fields introduces hidden documents as a consequence of the way ES indexes objects which in turn increases the document count in the stats. See [here](https://www.elastic.co/guide/en/elasticsearch/guide/current/nested-objects.html) for more detail.

File: data/mapping/mapping.json [hunk 2]
```diff
@@ -4,31 +4,25 @@
...
-            "dr",
-            "drs",
-            "partners",
-            "practice",
-            "surgery",
-            "medical",
-            "health",
-            "centre",
-            "center" ]
+          "stopwords": [ "doctor", "doctors", "dr", "drs" ]
```
Comment (neilbmclaughlin) [inline data/mapping/mapping.json:20]: Using common titles as stop words for GP names seems reasonable as there is no need to search by them in the same way as searches by surgery name may include common terms ('The Market Surgery' vs 'Dr Doctor').

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-22.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-22.dev.beta.nhschoices.net)

File: README.md [hunk 3]
```diff
@@ -4,39 +4,16 @@ Running [`scripts/start`](scripts/start) will start an Elasticsearch instance,
...
-    }
-  }
-}'
-```
+A basic set of tests can be run from [`scripts/test`](scripts/test)
 
-When running the container in this way, the port on which ES is exposed can be
-overridden by setting an environment variable - `ES_PORT` - equal to the port
-wishing to be used e.g. `ES_PORT=9201 ./scripts/start`.
+There are a number of sample queries in [sample-queries](./sample-queries) which can be used to query the data. (e.g  `curl -s -XPOST "http://localhost:9200/profiles/_search?pretty" -d @sample_queries/query-nested-doctor.json | jq -C '.' | less -R`)
```
Comment (st3v3nhunt) [inline README.md]: Can line breaks be added so no line is greater in length than 80 chars please?

File: README.md [hunk 4]
```diff
@@ -4,39 +4,16 @@ Running [`scripts/start`](scripts/start) will start an Elasticsearch instance,
...
 
 ## Optional
 
-Use [kibana](https://www.elastic.co/products/kibana) for ES querying and
-visualistion
+An alternative to using `curl` for ES config and querying is [kibana](https://www.elastic.co/products/kibana)
```
Comment (st3v3nhunt) [inline README.md]: Can line breaks be added so no line is greater in length than 80 chars please?

File: README.md [hunk 5]
```diff
@@ -4,39 +4,16 @@ Running [`scripts/start`](scripts/start) will start an Elasticsearch instance,
 loading data from
 [data/input/gp-data-merged.json](data/input/gp-data-merged.json) into an index
 named `profiles` [localhost:9200/profiles](http://localhost:9200/profiles).
+When running the container in this way, the port on which ES is exposed can be
+overridden by setting the environment variable `ES_PORT` (e.g. `ES_PORT=9201 ./scripts/start`).
```
Comment (st3v3nhunt) [inline README.md]: Can line breaks be added so no line is greater in length than 80 chars please?

File: data/scripts/transform-data-filter.txt [hunk 6]
```diff
@@ -0,0 +1,8 @@
+.[] | 
+.doctors = ( .doctors | map( { name: . } )) |
+.alternativeName = .address.addressLines[0] |
+.id = ._id |
+del (._id) |
+del (.openingTimes) |
+{"index": {"_index": "profiles", "_type": "gps", "_id": .id}}, .
+
```
Comment (st3v3nhunt) [inline data/scripts/transform-data-filter.txt]: Is the blank line needed?

File: data/scripts/transform-data-filter.txt [hunk 7]
```diff
@@ -0,0 +1,8 @@
+.[] |
```
Comment (st3v3nhunt) [inline data/scripts/transform-data-filter.txt]: Can this file's extension be renamed to `jq` to make it clear it is a `jq` filter or are there other consequences to doing that?

Also, there is a rogue space at the end of line 1.

File: data/mapping/mapping.json [hunk 8]
```diff
@@ -4,31 +4,25 @@
       "analyzer":{
         "gp_name_analyzer":{
           "type":"custom",
+          "tokenizer":"name_tokenizer",
+          "filter":[ "lowercase", "gp_title_stop" ]
+        },
+        "surgery_name_analyzer":{
```
Comment (st3v3nhunt) [inline data/mapping/mapping.json]: How is this different from a [standard analyser](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-standard-analyzer.html)?

File: data/mapping/mapping.json [hunk 9]
```diff
@@ -4,31 +4,25 @@
       "analyzer":{
         "gp_name_analyzer":{
```
Comment (st3v3nhunt) [inline data/mapping/mapping.json:5]: Could the `gp_name_analyzer` be a [stop analyzer](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-stop-analyzer.html)?

File: data/mapping/mapping.json (hunk 2 above)
Comment (st3v3nhunt) [reply data/mapping/mapping.json:20]: Seems fair enough to add possible alternatives, however, I've not seen anything other than `dr` in the name. Searched via:
* `jq '.[].doctors' data/input/gp-data-merged.json | grep <search_term>`
The only `<search_term>` returning results is `dr`.

Does the punctuation get removed from terms like `dr.` so just adding `dr` is going to cover it? The reason I ask is there 1625 results for `jq '.[].doctors' data/input/gp-data-merged.json | grep 'Dr\.' | wc -l`

File: data/mapping/mapping.json [hunk 10]
```diff
@@ -72,13 +65,12 @@
...
         },
         "doctors": {
-          "type": "text",
-          "analyzer": "gp_name_analyzer",
-          "fields": {
-            "keyword": {
-              "type": "keyword",
-              "ignore_above": 256
-            }
+          "type": "nested",
```
Comment (st3v3nhunt) [inline data/mapping/mapping.json:63]: I understand the need for setting `doctors` to be `nested` but I'm not sure the reason for the transformation in the first instance.

File: README.md (hunk 3 above)
Comment (st3v3nhunt) [inline README.md]: I'm a little unclear as to the point of the sample queries as they aren't all doing what `gp-finder` is doing. Can a `README` be added to the `sample-queries` directory explaining what each one is there for  and perhaps when each query might be useful to use?

**Review by st3v3nhunt — CHANGES_REQUESTED** [review]
Some niggly little changes to do please. Along with some explanation for why things have been done as my understanding most likely isn't quite correct.

**Comment by beseven** [comment]
@neilbmclaughlin will pull down this today and undoubtly will have some quesitons for you a bit later

File: data/mapping/mapping.json (hunk 2 above)
Comment (neilbmclaughlin) [reply data/mapping/mapping.json:20]: I think your grep needs to be case insensitive. `jq '.[].doctors' data/input/gp-data-merged.json | grep -i 'Doctor' ` gives me matches.

ES removes full stops and commas as well as a load of non-alphanumeric characters (~,@,etc) so the gp_name_tokenizer will split 'dr. andrew smith' into ['andrew', 'smith'].

ES has a cool _analyze API to show what the ouput for a given input would be:
```
curl -XGET 'localhost:9200/profiles/_analyze?pretty' -d'
{
  "analyzer" : "gp_name_analyzer",
  "text" : "Dr. Andrew Smith"
}'
```

File: data/scripts/transform-data-filter.txt (hunk 6 above)
Comment (neilbmclaughlin) [reply data/scripts/transform-data-filter.txt]: Removed

File: data/mapping/mapping.json (hunk 2 above)
Comment (st3v3nhunt) [reply data/mapping/mapping.json:20]: Yup, my bad omitted `-i`. Is it worth adding `gp` and possibly `general` and `practitioner`? That would cover this result `General practitioner Dr Bernadette Duffy `

File: data/mapping/mapping.json (hunk 9 above)
Comment (neilbmclaughlin) [reply data/mapping/mapping.json:5]: It's the custom name tokenizer which means it has to be a custom analyzer. The use of a custom tokenizer is preferred to the standard tokenizer as it handles searching for names like "O'Reilly" properly ([source](https://twitter.com/ErikRose/status/314135579266777089)).

File: data/mapping/mapping.json (hunk 8 above)
Comment (neilbmclaughlin) [reply data/mapping/mapping.json]: surgery_name_analyzer is no longer used. I've removed it.

File: data/mapping/mapping.json (hunk 10 above)
Comment (neilbmclaughlin) [reply data/mapping/mapping.json:63]: I think your grep needs to be case insensitive. `jq '.[].doctors' data/input/gp-data-merged.json | grep -i 'Doctor' ` gives me matches.

ES removes full stops and commas as well as a load of non-alphanumeric characters (~,@,etc) so the gp_name_tokenizer will split 'dr. andrew smith' into ['andrew', 'smith'].

ES has a cool _analyze API to show what the ouput for a given input would be:
```
curl -XGET 'localhost:9200/profiles/_analyze?pretty' -d'
{
  "analyzer" : "gp_name_analyzer",
  "text" : "Dr. Andrew Smith"
}'
```

Comment (neilbmclaughlin) [reply data/mapping/mapping.json:63]: Doctors needs to be nested as, with the existing string array of doctors names, a search for 'Andrew Smith' with find a match in [ 'Dr Jane Smith', 'Dr Andrew Jones'].

The problem is described [here](https://www.elastic.co/guide/en/elasticsearch/reference/current/nested.html).

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-22.dev.beta.nhschoices.net)

File: data/mapping/mapping.json (hunk 2 above)
Comment (neilbmclaughlin) [reply data/mapping/mapping.json:20]: I'm trying to stick to the clear definition of a stop word in that they are words which occur in so many documents as to make them useless as a search term. I'd argue that handling _general_ and _practitioner_ (as well as the various spelling permutations of professor) is probably a data cleansing exercise to create a proper 'title' property.

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-22.dev.beta.nhschoices.net)

**Review by st3v3nhunt — APPROVED** [review]
LGTM 👍

**Pull request merged**